### PR TITLE
Fixed lingering bug with setting the FSR Keys Map after reading in tracks

### PR DESCRIPTION
**Pull request by samuelshaner**
This PR fixed the lingering issue that we have had trouble debugging. The issue boiled down to the behavior of the methods used to get and set the `_FSR_keys_map` and `_FSRs_to_keys` variables in the Geometry. The methods have been updated to pass a reference to `_FSR_keys_map` and `_FSRs_to_keys` so the original object is modified instead of a new object being created and passed in to the geometry. 

Additionally, one bug was fixed in the ParallelHashMap where the locks needed to be unset in the clear() method. Also, several loops in the TrackGenerator have been parallelized.

This PR builds on PRs #270 and #284.

### CODE REVIEW

**Comment by wbinventor** [comment]
Holy cow, so it was a pass by reference issue?

**Comment by samuelshaner** [comment]
It was a pass by pointer issue. When the parallel hash map was passed as a pointer, it would get copied to the original parallel hash map associated with the Geometry. It doesn't appear that it was being copied over correctly. This was fixed by passing around the parallel hash map by reference and always accessing/modifying the original object.

**Comment by samuelshaner** [comment]
I should also give @geogunow a lot of the credit for his recommendations in helping figure out the cause and fix for this bug - thanks @geogunow!

**Comment by wbinventor** [comment]
I'll try to review this later today.

**Comment by samuelshaner** [comment]
Thanks @wbinventor. Just note that this builds on PR #270.

File: src/Geometry.cpp [hunk 1]
```diff
@@ -1113,8 +1112,8 @@ void Geometry::initializeCmfd() {
  * @brief Returns a pointer to the map that maps FSR keys to FSR IDs
  * @return pointer to _FSR_keys_map map of FSR keys to FSR IDs
  */
-ParallelHashMap<std::string, fsr_data*>* Geometry::getFSRKeysMap() {
-  return &_FSR_keys_map;
+ParallelHashMap<std::string, fsr_data*> & Geometry::getFSRKeysMap() {
```
Comment (wbinventor) [inline src/Geometry.cpp]: Is our new notation to use a blank space before `&`?

File: src/Geometry.cpp [hunk 2]
```diff
@@ -1122,33 +1121,8 @@ ParallelHashMap<std::string, fsr_data*>* Geometry::getFSRKeysMap() {
...
-}
-
-
-/**
- * @brief Sets the _FSRs_to_keys vector
- * @param FSRs_to_keys vector of FSR key hashes indexed by FSR IDs
- */
-void Geometry::setFSRsToKeys(std::vector<std::string>* FSRs_to_keys) {
-  _FSRs_to_keys = *FSRs_to_keys;
+std::vector<std::string> & Geometry::getFSRsToKeys() {
```
Comment (wbinventor) [inline src/Geometry.cpp]: Is our new notation to use a blank space before `&`?

File: src/Geometry.h [hunk 3]
```diff
@@ -148,18 +148,16 @@ class Geometry {
   void setRootUniverse(Universe* root_universe);
 
   Cmfd* getCmfd();
-  std::vector<std::string>* getFSRsToKeys();
+  std::vector<std::string> & getFSRsToKeys();
```
Comment (wbinventor) [inline src/Geometry.h]: Is our new notation to use a blank space before `&`?

File: src/Geometry.h [hunk 4]
```diff
@@ -148,18 +148,16 @@ class Geometry {
...
   int getFSRId(LocalCoords* coords);
   Point* getFSRPoint(int fsr_id);
   Point* getFSRCentroid(int fsr_id);
   std::string getFSRKey(LocalCoords* coords);
-  ParallelHashMap<std::string, fsr_data*>* getFSRKeysMap();
+  ParallelHashMap<std::string, fsr_data*> & getFSRKeysMap();
```
Comment (wbinventor) [inline src/Geometry.h]: Is our new notation to use a blank space before `&`?

File: src/ParallelHashMap.h [hunk 5]
```diff
@@ -566,23 +553,47 @@ void ParallelHashMap<K,V>::insert(K key, V value)
...
+ * @brief Updates the value associated with a key in the parallel hash map.
+ * @details The thread first acquires the lock for the bucket associated with
+ *      the key is acquired, then the linked list in the bucket is searched
+ *      for the key. If the key is not found, an exception is returned. When
+ *      the key is found, the value is updated and the lock is released.
+ * @param key the key of the key/value pair to be updated
+ * @param value the new value for the key/value pair
+ */
+template <class K, class V>
+void ParallelHashMap<K,V>::update(K key, V value) {
```
Comment (wbinventor) [inline src/ParallelHashMap.h]: Was this added in this PR? I was under the impression that @geogunow added this in #270.

File: src/Geometry.cpp (hunk 1 above)
Comment (samuelshaner) [inline src/Geometry.cpp]: Looks like the google c++ style guide recommends:

    Pointer operators have no space after the * or &.

So I'll remove the space before `&`.

File: src/Geometry.cpp (hunk 2 above)
Comment (samuelshaner) [inline src/Geometry.cpp]: ditto

File: src/ParallelHashMap.h (hunk 5 above)
Comment (samuelshaner) [inline src/ParallelHashMap.h]: @geogunow added this. Maybe I hadn't merged with his latest PR or something. I'll merge with develop and see if that changes anything.

File: src/TrackGenerator.cpp [hunk 6]
```diff
@@ -1434,12 +1453,11 @@ void TrackGenerator::dumpTracksToFile() {
   }
 
   /* Get FSR vector maps */
-  ParallelHashMap<std::string, fsr_data*>* FSR_keys_map =
+  ParallelHashMap<std::string, fsr_data*> & FSR_keys_map =
```
Comment (wbinventor) [inline src/TrackGenerator.cpp]: ditto space before `&`

File: src/TrackGenerator.cpp [hunk 7]
```diff
@@ -1434,12 +1453,11 @@ void TrackGenerator::dumpTracksToFile() {
...
       _geometry->getFSRKeysMap();
-  std::vector<std::string>* FSRs_to_keys = _geometry->getFSRsToKeys();
+  std::vector<std::string> & FSRs_to_keys = _geometry->getFSRsToKeys();
```
Comment (wbinventor) [inline src/TrackGenerator.cpp]: ditto space before `&`

**Comment by wbinventor** [comment]
This looks good and worked well for me when running the SPH problems. I'll merge it in once Travis finishes.

**Pull request merged**